lib/comgr/cmake/opencl_pch.cmake

```diff
@@ -1,26 +1,8 @@
 if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
   find_package(Clang REQUIRED CONFIG)
-
-  # FIXME: CLANG_CMAKE_DIR seems like the most stable way to find this, but
-  # really there is no way to reliably discover this header.
-  #
-  # We effectively back up to the Clang output directory (for the case of a build
-  # tree) or install prefix (for the case of an installed copy), and then search
-  # for a file named opencl-c.h anywhere below that. We take the first result in
-  # the case where there are multiple (e.g. if there is an installed copy nested
-  # in a build directory). This is a bit imprecise, but it covers cases like MSVC
-  # adding some additional configuration-specific subdirectories to the build
-  # tree but not to an installed copy.
-  file(GLOB_RECURSE OPENCL_C_H_LIST "${CLANG_CMAKE_DIR}/../../../*/opencl-c.h")
-
-  list(GET OPENCL_C_H_LIST 0 OPENCL_C_H)
-
-  if (NOT EXISTS "${OPENCL_C_H}" OR IS_DIRECTORY "${OPENCL_C_H}")
-    message(FATAL_ERROR "Unable to locate opencl-c.h from the supplied Clang. The path '${CLANG_CMAKE_DIR}/../../../*' was searched.")
-  endif()
-else()
-  get_target_property(clang_build_header_dir clang-resource-headers RUNTIME_OUTPUT_DIRECTORY)
-  set(OPENCL_C_H "${clang_build_header_dir}/opencl-c.h")
+  execute_process(COMMAND "${CLANG_INSTALL_PREFIX}/bin/clang" -print-resource-dir OUTPUT_VARIABLE CLANG_RESOURCE_DIR)
```

Comment:
Does this patch work if clang is only built but not installed? The original method works even if clang is not installed.

Reply:
Can this simply refer to the `clang` target directly? I don't think we should need to find the actual clang executable ourselves.

I believe you can use https://cmake.org/cmake/help/latest/command/add_custom_command.html as in `add_custom_command(COMMAND clang` and it will manage finding the appropriate `LOCATION` for the `clang` executable target. For some reason it doesn't support `OUTPUT_VARIABLE` though. However this means it will happen at build-time rather than configure-time; I am not actually certain which is appropriate here? I would think we want this to run whenever the underlying `clang` changes, which could happen without a re-configure, right? In that case, we actually want the build-time behavior anyway.

Alternatively, you can do the `LOCATION` lookup from the `clang` target yourself, and still use `execute_process`. In that case I think the only option we have is configure-time.

Reply:
> Does this patch work if clang is only built but not installed? The original method works even if clang is not installed.

It can work if clang not installed. I uninstalled clang in system, and build clang at `<ClangBuild>`, then `export Clang_DIR="<ClangBuild>/cmake/modules/CMakeFiles/"` (which contains the correct `${CLANG_INSTALL_PREFIX}`) . Once it finds the clang executable, `-print-resource-dir` will work, even clang is just built, not installed. But I find `<ClangBuild>/cmake/modules/CMakeFiles/ClangConfig.cmake` causes issues at build time, because it does not specify the correct `CLANG_INCLUDE_DIRS`, while `<ClangBuild>/lib64/cmake/clang/ClangConfig.cmake` provides the correct `CLANG_INCLUDE_DIRS` but did not provide `CLANG_INSTALL_PREFIX`

So I suppose the best approach may be using `${CLANG_CMAKE_DIR}/../../../bin/clang -print-resource-dir`

Reply:
> I believe you can use https://cmake.org/cmake/help/latest/command/add_custom_command.html as in add_custom_command(COMMAND clang and it will manage finding the appropriate LOCATION for the clang executable target.

Does `add_custom_command(COMMAND clang` uses clang exe in `PATH`, or cmake find the clang command according to clang's cmake module? If it's the previous case, the command will fail if clang not installed, just built at a custom directory.

Reply:
I switched to using `${CLANG_CMAKE_DIR}/../../../bin/clang -print-resource-dir`, because for installed clang, `CLANG_CMAKE_DIR` is `${CLANG_INSTALL_PREFIX}/lib64/cmake/clang`, defined in `ClangConfig.cmake`; while for build-only clang, `CLANG_CMAKE_DIR` is hardcoded into `<clang-build-dir>/lib64/cmake/clang/ClangConfig.cmake` so its also reliable.

Reply:
BTW I'm using the `${CLANG_CMAKE__DIR}` which is used before this patch, so this patch would not bring more vulnerability.